### ISSUE TITLE
Update <experimental/optional> to implement LFTS v2.

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#30419]
|  |  |
| --- | --- |
| Bugzilla Link | [31071](https://llvm.org/bz31071) |
| Version | unspecified |
| OS | MacOS X |
| Attachments | [Sample project](https://user-images.githubusercontent.com/60944935/143754724-85cafc02-2bcf-4166-b83e-791e7866bd93.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@mclow,@zygoloid |

## Extended Description 

Suppose we a function that takes a completion handler in the form of an optional<function>:

    using Handler = function<void(void)>;
    
    void doSomething(int flags, optional<Handler> callback = optional<Handler>()) {
        if (callback) {
            (*callback)();
        }
    }

Since a default value of an empty optional is provided, I can invoke this without passing a value for the final parameter just fine:

    doSomething(0);

Moreover, I can invoke `doSomething` if I explicitly cast a lambda expression to a `Handler` and pass it for the final parameter:

    doSomething(0, Handler([]{
        cout << "hello world" << endl;
    }));

However, passing a lambda expression directly does not work (the following does not compile with "No matching function for call to 'doSomething'"):

    doSomething(0, []{
       cout << "hello world" << endl;
    });

I assume this is because the type checker doesn't know we can assign a lambda expression to an optional (even if it's an optional<function> and lambda expression can be assigned to functions).

In other words, if a lambda expression L can be assigned to a function F, we should be able to assign L to an optional of type optional<F>.

A sample project that demonstrates this is attached. I am testing with Xcode 8.1 GM.

**Comment by philnik777** [comment]
`<experimental/optional>` has been removed.
